File: room-manager/src/main/java/org/rivero/roommanager/controllers/UserController.java
package org.rivero.roommanager.controllers;

import lombok.RequiredArgsConstructor;
import org.rivero.roommanager.entities.User;
import org.rivero.roommanager.repositories.UserRepository;
import org.rivero.roommanager.request.LoginRequest;
import org.rivero.roommanager.request.UserUpdateRequest;
import org.rivero.roommanager.services.UserService;
import org.rivero.roommanager.user.MetadataItem;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Flux;

import java.util.List;

@RestController
@RequestMapping("/api/v1")
@RequiredArgsConstructor
public class UserController {
    private final UserService userService;
    private final UserRepository userRepository;

    @GetMapping("/members")
    public Flux<MetadataItem> getMembers() {
        return Flux.fromStream(userRepository.findAll().stream()
                .map(u -> new MetadataItem(u.getId(), u.getName())));
    }

    @GetMapping("/users")
    public List<User> getUser() {
        return userService.getAllUser();
    }

    @PostMapping("/users")
    public ResponseEntity<User> addUser(@RequestBody User user) {
        userService.register(user);
        return ResponseEntity.ok(user);
    }

    @PostMapping("/login")
    public ResponseEntity<String> login(@RequestBody LoginRequest request) {
        return ResponseEntity.ok(userService.login(request));
    }

    @GetMapping("/users/{userId}")
    public ResponseEntity<User> getUserById(@PathVariable(name = "userId") String userId) {
        if (userService.getUserById(userId) != null)
            return ResponseEntity.ok(userService.getUserById(userId));
        return ResponseEntity.notFound().build();
    }

    @PutMapping("/user/{userId}")
    public ResponseEntity<String> updateUserById(@PathVariable(name = "userId") String userId, @RequestBody UserUpdateRequest user) {
        userService.updateOne(user, userId);
        return ResponseEntity.ok().build();
    }

    @DeleteMapping("/user/{userId}")
    public ResponseEntity<Void> deleteUserById(@PathVariable(name = "userId") String userId) {
        userService.deleteOne(userId);
        return ResponseEntity.noContent().build();
    }
}

File: room-manager/src/main/kotlin/org/rivero/roommanager/user/UserController.kt
package org.rivero.roommanager.user

import lombok.RequiredArgsConstructor
import org.rivero.roommanager.entities.User
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.*
import org.springframework.web.server.ResponseStatusException
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/api/v1")
@RequiredArgsConstructor
class UserController @Autowired constructor(
    val userService: UserService,
    val userRepository: UserRepository
) {

    @GetMapping("/members")
    fun getMembers(): Flux<MetadataItem> {
        return Flux.fromStream(userRepository.findAll().stream())
            .map { it!! }
            .map { u -> MetadataItem(u.id, u.name) }
    }

    @GetMapping("/users")
    fun getUsers(): Flux<User> {
        return Flux.fromStream(userRepository.findAll().stream());
    }


    @PostMapping("/register")
    @ResponseStatus(HttpStatus.CREATED)
    fun register(): Mono<Any> {
        return userService.register();
    }

    @GetMapping("/users/{id}")
    fun getUserById(@PathVariable(name = "id") id: String): Mono<UserDto> {
        return Mono.justOrEmpty(userRepository.findById(id))
            .switchIfEmpty(Mono.error(ResponseStatusException(HttpStatus.NOT_FOUND, "User not found")))
            .map { it!! }
            .map { u -> UserDto(u.id, u.name) }
    }

    @PutMapping("/users/{id}")
    @ResponseStatus(HttpStatus.OK)
    fun updateUser(
        @PathVariable(name = "id") id: String,
        @RequestBody request: UpdateUserRequest
    ): Mono<Any> {
        return Mono.just(
            userService.update(id, request)
        )
    }

    @DeleteMapping("/users/{id}")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteUserById(@PathVariable(name = "id") id: String) {
        userService.deleteOne(id)
    }
}
